File: Graph/DFS.java
public class DFS {
    private ArrayList<ArrayList<Integer>> graph;
    private boolean[] visited;
    private ArrayList<Integer> traversalOrder;

    public DFS(int vertices) {
        graph = new ArrayList<>();
        for (int i = 0; i < vertices; i++) {
            graph.add(new ArrayList<>());
        }
    }

    // 양방향 간선 추가
    public void addEdge(int u, int v) {
        graph.get(u).add(v);
        graph.get(v).add(u);
    }

    // 단방향 간선 추가
    public void addDirectedEdge(int from, int to) {
        graph.get(from).add(to);
    }

    // 모든 vertex에 대한 DFS
    public ArrayList<Integer> traverseAll() {
        visited = new boolean[graph.size()];
        traversalOrder = new ArrayList<>();
        
        for (int vertex = 0; vertex < graph.size(); vertex++) {
            if (!visited[vertex]) {
                dfsRecursive(vertex);
            }
        }
        
        return traversalOrder;
    }

    // 특정 vertext부터 DFS
    public ArrayList<Integer> traverseFrom(int startVertex) {
        visited = new boolean[graph.size()];
        traversalOrder = new ArrayList<>();
        dfsRecursive(startVertex);
        return traversalOrder;
    }

    private void dfsRecursive(int currentVertex) {
        visited[currentVertex] = true;
        traversalOrder.add(currentVertex);
        
        for (int neighbor : graph.get(currentVertex)) {
            if (!visited[neighbor]) {
                dfsRecursive(neighbor);
            }
        }
    }
}

File: Graph/DFS.kt
class DFS(vertices: Int) {
    private val graph: MutableList<MutableList<Int>> = MutableList(vertices) { mutableListOf() }
    private lateinit var visited: BooleanArray
    private lateinit var traversalOrder: MutableList<Int>
    
    // 양방향 간선 추가
    fun addEdge(u: Int, v: Int) {
        graph[u].add(v)
        graph[v].add(u)
    }
    
    // 단방향 간선 추가
    fun addDirectedEdge(from: Int, to: Int) {
        graph[from].add(to)
    }
    
    // 모든 vertex에 대한 DFS
    fun traverseAll(): List<Int> {
        visited = BooleanArray(graph.size)
        traversalOrder = mutableListOf()
        
        for (vertex in graph.indices) {
            if (!visited[vertex]) {
                dfsRecursive(vertex)
            }
        }
        
        return traversalOrder
    }
    
    // 특정 vertex부터 DFS
    fun traverseFrom(startVertex: Int): List<Int> {
        visited = BooleanArray(graph.size)
        traversalOrder = mutableListOf()
        dfsRecursive(startVertex)
        return traversalOrder
    }
    
    private fun dfsRecursive(currentVertex: Int) {
        visited[currentVertex] = true
        traversalOrder.add(currentVertex)
        
        for (neighbor in graph[currentVertex]) {
            if (!visited[neighbor]) {
                dfsRecursive(neighbor)
            }
        }
    }
}
